### PR TITLE
New treestate

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Move `MonadProtocolVersion` into base.

This supports the new consensus protocol.

## Changes

- Moved `MonadProtocolVersion` to `Concordium.Types`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: haskell-src/Concordium/Genesis/Data/BaseV1.hs [hunk 1]
```diff
@@ -58,3 +65,25 @@ instance Serialize RegenesisDataV1 where
         genesisTerminalBlock <- get
         genesisStateHash <- get
         return RegenesisDataV1{..}
+
+-- |Convert 'GenesisParametersV2' to genesis data.
```
Comment (EmilLa1) [inline haskell-src/Concordium/Genesis/Data/BaseV1.hs:69]:
```suggestion
-- |Convert 'GenesisParametersV3' to genesis data.
```

File: haskell-src/Concordium/Genesis/Data/BaseV1.hs [hunk 2]
```diff
@@ -58,3 +65,25 @@ instance Serialize RegenesisDataV1 where
...
+    genesisAccounts = Vec.fromList gp3InitialAccounts
+    genesisCryptographicParameters = gp3CryptographicParameters
+    genesisIdentityProviders =
+        case filter (\(k, v) -> k /= ipIdentity v) (Map.toList (idProviders gp3IdentityProviders)) of
+            [] -> gp3IdentityProviders
+            ips -> error $ "Inconsistent identity provider ids: " ++ show ips
+    genesisAnonymityRevokers =
+        case filter (\(k, v) -> k /= arIdentity v) (Map.toList (arRevokers gp3AnonymityRevokers)) of
+            [] -> gp3AnonymityRevokers
+            ars -> error $ "Inconsistent anonymity revoker ids: " ++ show ars
```
Comment (abizjak) [inline haskell-src/Concordium/Genesis/Data/BaseV1.hs:87]: I think I had this comment in a previous PR already, but checking these seems arbitrary here.

For example we don't check that the anonymity revokers and identity providers that appear in the accounts are the right ones either.

So I think checking this is arbitrary and unclear why we need to bother.

File: haskell-src/Concordium/Genesis/Data/P6.hs [hunk 3]
```diff
@@ -73,6 +74,9 @@ putVersionedGenesisData gd = do
     putVersion 8
     putGenesisDataV8 gd
 
+parametersToGenesisData :: GenesisParametersV3 'P6 -> GenesisDataP6
+parametersToGenesisData = uncurry GDP6Initial . BaseV1.parametersToState
```
Comment (abizjak) [inline haskell-src/Concordium/Genesis/Data/P6.hs:78]: In a previous PR this function was removed with the argument that it is not needed. Why is it being reintroduced?

Comment (td202) [reply haskell-src/Concordium/Genesis/Data/P6.hs]: This was an oversight as this branched from an old version. `GenesisParametersV3` has now been (re)removed.

**Review by td202 — APPROVED** [review]

File: haskell-src/Concordium/Types.hs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE ConstraintKinds #-}
```
Comment (abizjak) [inline haskell-src/Concordium/Types.hs:1]: this does not seem to be needed

**Review by abizjak — APPROVED** [review]

File: haskell-src/Concordium/Types.hs (hunk 4 above)
Comment (EmilLa1) [reply haskell-src/Concordium/Types.hs]: Nice catch!

**Review by limemloh — APPROVED** [review]

**Pull request merged**